Sources/_NIOConcurrency/AsyncSequenceSupport.swift

```diff
@@ -0,0 +1,141 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if swift(>=5.5)
+import struct NIO.ByteBuffer
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+public struct ByteBufferToUInt8AsyncSequence<Upstream: AsyncSequence>: AsyncSequence where Upstream.Element == ByteBuffer {
+    public typealias Element = UInt8
+    public typealias AsyncIterator = Iterator
+    
+    public struct Iterator: AsyncIteratorProtocol {
+        /*private but*/ @usableFromInline var state: State
+        
+        @usableFromInline
+        enum State {
+            case hasBuffer(ByteBuffer, Upstream.AsyncIterator)
+            case askForMore(Upstream.AsyncIterator)
+            case finished
+            case modifying
+        }
+        
+        @usableFromInline
+        init(_ upstream: Upstream.AsyncIterator) {
+            self.state = .askForMore(upstream)
+        }
+        
+        @inlinable
+        public mutating func next() async throws -> Element? {
+            switch self.state {
+            case .askForMore(var upstream):
+                self.state = .modifying
+                
+                while true {
+                    switch try await upstream.next() {
+                    case .some(let nextBuffer) where nextBuffer.readableBytes == 0:
+                        break
```

Comment:
Please avoid `break` in `switch` statements within loops: it can be very hard to work out what this is doing. I suggest either using a labelled loop and then a labelled `continue` statement, or just writing `()` in the body.